コメント投稿機能を追加

mterada1228 opened the pull request.

mterada1228 committed:
comments テーブルを作成する migration ファイルを作成する

mterada1228 committed:
Comment モデルを追加

mterada1228 committed:
コメント投稿機能の画面モックを作成

mterada1228 committed:
title カラムは不要なので削除する

mterada1228 committed:
コメントを投稿し、テーブルに保存する

mterada1228 committed:
コメント投稿機能を実装

mterada1228 commented on app/views/races/horses/comments/index.html.erb:
```diff
@@ -1,8 +1,41 @@
+          <%= f.text_field :user_name, class: 'form-control', placeholder: 'ニックネームを入力' %>
+        </div>
+        <div class="form-group comment-header-area">
+          <%= f.select :horse_id, [['ウインエアフォルク', 0], ['アーモンドアイ', 1], ['ゴールドシップ', 2]], {}, { class: 'custom-select' } %>
```
馬名選択機能は、レースに出走する予定の馬から選択できるよう、別 PR で対応予定です。

mterada1228 commented on app/views/races/_race_info.html.erb:
```diff
@@ -0,0 +1,27 @@
+      </div>
+    </div>
+  </div>
+</div>
```
こちらの Partial はコメント投稿には直接関係ないので現状モックです。  
別 PR でレース情報を動的に取得できるよう修正予定です。

mterada1228 committed:
ファイルの末尾に改行を加える

mterada1228 commented on app/views/races/_race_info.html.erb:
```diff
@@ -0,0 +1,27 @@
+<div class="container">
```
こちらの Partial はコメント投稿には直接関係ないので、現状はモックです。  
別 PR にて動的にレース情報を取得できるよう修正予定です。

ershn commented on app/models/comment.rb:
```diff
@@ -0,0 +1,9 @@
+class Comment < ApplicationRecord
+  belongs_to :horse
+  belongs_to :race
```
`horse_race`と紐付けた方が柔軟なデータモデルになると思いました。

mterada1228 replied:
`comment` は `HorseRace` に紐づくよう修正しました。

ershn replied:
OK

ershn commented on app/controllers/races/horses/comments_controller.rb:
```diff
@@ -2,8 +2,25 @@ module Races
+      private
+
+      def comment_params
+        params.require(:comment).permit(:horse_id, :race_id, :description, :user_name, :position)
```
バリデーションなしでレコード id を受け取ると、パラメーターの改ざんで自由に設定できるようになりますが、それだと、本来設定できないはずのものが設定可能になります。

ここはバリデーションを追加する必要があります。

mterada1228 replied:
`Comment` を `HorseRace` に `belongs_to` させるように修正したため、`horse_id` と `race_id` を受け取る必要は無くなりました。

代わりに存在しない、`horse_race_id` を受け取った時は、404 エラーとしてエラーページを表示するような `before_action` を実装しています。

ershn replied:
OK

ershn commented on app/controllers/races/horses/comments_controller.rb:
```diff
@@ -2,8 +2,25 @@ module Races
+      private
+
+      def comment_params
+        params.require(:comment).permit(:horse_id, :race_id, :description, :user_name, :position)
```
url のパスで受け取るものはパラメーターで設定できないようにした方が安全です。

mterada1228 replied:
`horse_id`, `race_id` は、`Comment` が `HorseRace` に `belongs_to` するようになったため不要になりました。

代わりに `horse_race_id` が必要となりましたが、パラメータではなく URL からのみ受け取れるようにしています。

ershn replied:
OK

mterada1228 committed:
マイグレーションファイルの名前がモデルと重複していてエラーとなっていたので修正

ershn commented on config/routes.rb:
```diff
@@ -3,7 +3,7 @@
   resources :races, only: [:index, :show] do
     resources :horses, only: [:show] do
-      resources :comments, only: [:index], controller: 'races/horses/comments'
+      resources :comments, only: [:index, :create], controller: 'races/horses/comments'
```
`controller`の指定は必要ないはずなので、rails のルールを確認しておいた方がいいと思います。

mterada1228 replied:
`controllers` 配下に置くことで、指定の必要が無くなりました。

ershn replied:
OK

ershn commented on db/migrate/20210710060250_create_comment.rb:
```diff
@@ -0,0 +1,17 @@
+      t.string :race_id, null: false
+      t.string :title, null: false
+      t.text :description, null: false
+      t.string :user_name
```
これだと、ユーザー名は空文字と null の両方がありえますが、そのどちらかに合わせた方がデータが扱いやすくなります。

空白文字しかない、もしくは空文字のユーザー名は null に統一しておくのがいいと思います。

mterada1228 replied:
`Comment` の `before_save` で、空文字の場合は `nil` に変換するように修正しました。

ershn replied:
OK

ershn commented on spec/factories/comments.rb:
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :comment do
+    sequence(:id)
```
こちらは必要ないと思います。

mterada1228 replied:
こちら削除しました。

ershn replied:
OK

ershn commented on spec/features/races/horses/comments_spec.rb:
```diff
@@ -17,6 +18,59 @@
+          expect(page).to have_text('コメントの投稿に失敗しました。')
+          expect(page).to have_text('コメントは999文字以内で入力してください')
+        end
+      end
```
ポジションのテストも追加しておいたほうが安全だと思います。

mterada1228 replied:
こちら追加いたしました。

ershn replied:
OK

ershn commented on spec/models/comment_spec.rb:
```diff
@@ -0,0 +1,42 @@
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
```
`require`も`type: :model`も必要ないと思います。

mterada1228 replied:
こちら削除いたしました。

ershn replied:
OK

ershn commented on spec/models/comment_spec.rb:
```diff
@@ -0,0 +1,42 @@
+      end
+    end
+  end
+end
```
正常系のテストも追加しておいたほうが安全です。

mterada1228 replied:
正常系のテストを追加しました。

ershn replied:
OK

mterada1228 committed:
Comment は、horse, race ではなく horse_race に関連するようにする

mterada1228 committed:
HorseRace が存在しない場合は、404エラーとする

mterada1228 committed:
Comment 保存時に、user name が空文字の場合は nil に変換する

mterada1228 committed:
RSpec の修正

mterada1228 commented on config/database.yml:
```diff
@@ -1,7 +1,7 @@
   adapter: mysql2
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
-  encoding: utf8
+  encoding: utf8mb4
```
使用している MySql のエンコーディングに合致していないことを発見したため、こちら修正しました。

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,31 @@
+    @horse_race = HorseRace.find(params[:horse_race_id])
+  rescue ActiveRecord::RecordNotFound => _e
+    render template: 'errors/error_404', status: 404
+  end
```
create と set_horse_race の間でのその処理の分割だと、メソッド間でステートへの依存が生まれて、密結合になってしまいますので、避けた方がいいです。

ここは以下のように疎結合にしたほうが保守性が上がります。

CommentsController

```ruby
def create
  comment = HorseRace.find(params[:horse_race_id]).comments.new(comment_params)
  ...
```

ApplicationController

```ruby
rescue_from ActiveRecord::RecordNotFound do ||
  ...
end
```

mterada1228 replied:
ご指摘の通り修正いたしました。

ershn replied:
OK

ershn commented on spec/factories/comments.rb:
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :comment do
+    description { Faker::String.random(length: 999) }
+    user_name { Faker::Name.name }
```
こちらは 29 文字を超えないようにしたほうが安全かと。

mterada1228 replied:
29 文字で生成されるよう修正しました。

ershn replied:
OK

ershn commented on spec/models/comment_spec.rb:
```diff
@@ -0,0 +1,71 @@
+
+  describe 'validation' do
+    context '正常系' do
+      let(:comment) { create(:comment, horse_race: horse_race) }
```
build にした方が適切だと思います。

mterada1228 replied:
ご指摘の通り修正いたしました。

ershn replied:
OK

ershn commented on app/models/comment.rb:
```diff
@@ -0,0 +1,16 @@
+
+  validates :description, presence: true, length: { maximum: 999 }
+  validates :user_name, length: { maximum: 29 }
+
```
position のバリデーションを追加する必要があるかもしれません。

mterada1228 replied:
enum に 存在しない値を指定して、ActiveRecord を initializeしようとすると、validation より前に、ArgumentError が raise されてしまいます。

なので、今のところ2案考えているのですが、ベターなのはどちらか相談させてください。

1. `CommentsController` で `ArgumentError` を `rescue` してエラーハンドリング（500ページに飛ばす？）
1.  `ActiveRecord::Enum::EnumType#assert_valid_value` をオーバライドしてエラーを raise しないようにする。
その上で `Comments` モデルに `position` のバリデーションを追加する。
cf) https://github.com/rails/rails/issues/13971#issuecomment-383905807

ershn replied:
いい解決策が今のところ思いつかないので、issue として一旦お預かりにしましょう。

mterada1228 committed:
指摘対応: CommentsController#create と CommentsController#set_horse_race の密結合をやめる
  - HorseRace は CommentsController#create 内で取得するようにする
  - rescue は rescue_from で処理を分離するようにする

mterada1228 committed:
指摘対応: テスト修正
  - Comment の Factory は user_name を 29 文字以内で生成するようにする
  - Comment の model スペックは正常系でモデルを create ではなく build で生成する

ershn commented on spec/factories/comments.rb:
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :comment do
+    description { Faker::String.random(length: 999) }
+    user_name { Faker::Name.initials(number: 29) }
```
以上の２つで文字数は固定となりますが、それだと、一つのケースしかテストされなくなるので、カラムが許容する範囲でできるだけランダムな値を生成した方がいいと思います。

---

Faker が提供するオプションで文字数を制御できない場合は ruby のメソッドを使う手もあります。

`Faker::Name.name[0,29]`

mterada1228 committed:
指摘対応

mterada1228 committed:
Merge branch 'develop' into features/post_comments
